docs/tutorials/notebooks/unbalanced_optimal_transport.ipynb

```diff
@@ -0,0 +1,726 @@
+{
```

Comment:
<p>This looks like it really blew up, were you able to fix it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/289/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Reply:
<p>Unfortunately not yet, which is a bit surprising as training was really stable for our paper. In fact, there are two  difference, which is that we trained with early stopping and best model performance - and the <code>amortization_loss</code></p><p>Then I tried running it with <code>amortization_loss=regression</code> (the fully balanced case, so code changes should not have changed, but I can check again on the original branch), and it threw an error. </p><pre class="ql-syntax" spellcheck="false">TypeError                                 Traceback (most recent call last)
Cell In[16], line 41
     25         plt.close(fig)
     28 neural_dual_solver = neuraldual.W2NeuralDual(
     29     input_dim,
     30     neural_f,
   (...)
     39     #epsilon=1,
     40 )
---&gt; 41 learned_potentials = neural_dual_solver(
     42     *train_dataloaders,
     43     *valid_dataloaders,
     44     callback=training_callback,
     45 )
     46 clear_output()

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:255, in W2NeuralDual.__call__(self, trainloader_source, trainloader_target, validloader_source, validloader_target, callback)
    246 def __call__(
    247     self,
    248     trainloader_source: Iterable[jnp.ndarray],
   (...)
    253 ) -&gt; Union[potentials.DualPotentials, Tuple[potentials.DualPotentials,
    254                                             Train_t]]:
--&gt; 255   logs = self.train_fn(
    256       trainloader_source,
    257       trainloader_target,
    258       validloader_source,
    259       validloader_target,
    260       callback=callback,
    261   )
    262   res = self.to_dual_potentials()
    264   return (res, logs) if self.logging else res

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:377, in W2NeuralDual.train_neuraldual_alternating(self, trainloader_source, trainloader_target, validloader_source, validloader_target, callback)
    374   batch_g["source"] = jnp.asarray(next(trainloader_source))
    375   batch_g["target"] = jnp.asarray(next(trainloader_target))
--&gt; 377   self.state_g, loss_g, _ = self.train_step_g(
    378       self.state_f, self.state_g, batch_g
    379   )
    381 # get train batch for potential f
    382 batch_f["source"] = jnp.asarray(next(trainloader_source))

    [... skipping hidden 12 frame]

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:502, in W2NeuralDual.get_step_fn.&lt;locals&gt;.step_fn(state_f, state_g, batch)
    499 grad_fn = jax.value_and_grad(loss_fn, argnums=[0, 1], has_aux=True)
    500 if train:
    501   # compute loss and gradients
--&gt; 502   (loss, (loss_f, loss_g, W2_dist)), (grads_f, grads_g) = grad_fn(
    503       state_f.params,
    504       state_g.params,
    505       state_f.potential_value_fn,
    506       state_g.potential_value_fn,
    507       state_g.potential_gradient_fn,
    508       batch,
    509   )
    511   # update state
    512   if to_optimize == "both":

    [... skipping hidden 8 frame]

File ~/git_repos/ott-1/src/ott/solvers/nn/neuraldual.py:464, in W2NeuralDual.get_step_fn.&lt;locals&gt;.loss_fn(params_f, params_g, f_value, g_value, g_gradient, batch)
    459 elif self.amortization_loss == 'objective':
    460   f_value_parameters_detached = lambda x: f_value(
    461       jax.lax.stop_gradient(params_f), x
    462   )
    463   amor_loss = (
--&gt; 464       f_value_parameters_detached(init_source_hat) -
    465       batch_dot(init_source_hat, target)
    466   ).mean()
    467 else:
    468   raise ValueError("Amortization loss has been misspecified.")

TypeError: unsupported operand type(s) for -: 'function' and 'JVPTracer' 
</pre>

Reply:
Ah sorry about that, maybe we should update the neural dual test to sweep over `amortization_loss` in "regression" and "objective". That error can be fixed by updating that line of `neuraldual` to:

```python
elif self.amortization_loss == 'objective':
  f_value_parameters_detached = f_value(
      jax.lax.stop_gradient(params_f), g_value_partial)
  amor_loss = (
      f_value_parameters_detached(init_source_hat) -
      batch_dot(init_source_hat, target)
  ).mean()
```

Can you push this update into this PR? I could also make another small PR with this but it seems a little simpler to keep it in here

Reply:
With the objective loss your example still seems to be blowing up. I think it could be related to the ICNN architecture as I wasn't able to find an extremely stable/optimal one. If you have another architecture from your paper it may be worth trying that instead of the default OTT one. I've also found the MLP potentials to work well and be stable in a lot of cases.

![image](https://user-images.githubusercontent.com/707462/219480208-97a04f7a-6a90-4221-a7d2-0edcafcbb5dd.png)

Reply:
Oh I think I did not use the objective loss here, sorry, the order of execution is not yet correct, as I just did this push now to show the small bug. Thanks for letting me know how to fix it.

We actually used the same architecture, i.e. with [64,64,64,64] layers. I'll give it another try tomorrow!

Reply:
Maybe we should make this clearer somewhere but you can also disable the conjugate fine-tuning by setting `conjugate_solver=None` in case that's messing up the convergence